ksqldb-engine/src/main/java/io/confluent/ksql/structured/SchemaKStream.java

```diff
@@ -306,20 +306,31 @@ static Expression rewriteTimeComparisonForFilter(final Expression expression) {
 
   @SuppressWarnings("unchecked")
   public SchemaKStream<Struct> selectKey(
+      final FormatInfo valueFormat,
       final Expression keyExpression,
       final Optional<FormatInfo> forceInternalKeyFormat,
-      final Stacker contextStacker
+      final Stacker contextStacker,
+      final boolean forceRepartition
   ) {
     final boolean keyFormatChange = forceInternalKeyFormat.isPresent()
         && !forceInternalKeyFormat.get().equals(keyFormat.getFormatInfo());
 
-    if (!keyFormatChange && repartitionNotNeeded(ImmutableList.of(keyExpression))) {
+    if (!keyFormatChange
+        && repartitionNotNeeded(ImmutableList.of(keyExpression))
+        && !forceRepartition
+    ) {
       return (SchemaKStream<Struct>) this;
     }
 
     if (keyFormat.isWindowed()) {
-      throw new KsqlException("Implicit repartitioning of windowed sources is not supported. "
-          + "See https://github.com/confluentinc/ksql/issues/4385.");
+      final String errorMsg = "Implicit repartitioning of windowed sources is not supported. "
+          + "See https://github.com/confluentinc/ksql/issues/4385.";
+      final String additionalMsg = forceRepartition
+          ? " As a result, ksqlDB does not support joins on Schema-Registry-enabled key formats "
```

Comment:
This PR does not add support for joining on windowed sources with SR-enableld key formats. This will be a follow-up PR.